feat: Introduce NodePool repair policies with toleration durations

matanbaruch opened the pull request:
This pull request introduces a new feature that allows users to configure custom node repair toleration durations on a per-NodePool basis. Previously, node repair timing was dictated solely by CloudProvider defaults, but now users can override these defaults for specific node conditions or set a NodePool-wide default. This enables more granular control to match the needs of different workloads (e.g., stateful vs. stateless applications). The changes include updates to the NodePool API, CRDs, deep copy methods, documentation, and controller logic.

**NodePool Repair Configuration Feature:**

* Added a new `repair` block to the `NodePoolSpec` API, allowing users to specify custom repair policies and default toleration durations for node conditions. This includes the `RepairSpec` and `NodePoolRepairPolicy` types. [[1]](diffhunk://#diff-c1b722216cedf83c042b577c91a3d890a216d93f0fd5a85909683ae3b3365bd6R69-R74) [[2]](diffhunk://#diff-c1b722216cedf83c042b577c91a3d890a216d93f0fd5a85909683ae3b3365bd6R169-R208)
* Updated the CRDs (`karpenter.sh_nodepools.yaml`) to support the new `repair` configuration, including validation patterns and documentation for the new fields. [[1]](diffhunk://#diff-542ac25ed1857604758c133b5f49049a6ffac641ddd6202c3a34cb528f196a6eR175-R219) [[2]](diffhunk://#diff-4e0265201c4542984a38dca2ae7a5311bc50df00e0bd483148222b92726bfa37R175-R219)
* Implemented deep copy methods for the new `RepairSpec` and `NodePoolRepairPolicy` types to ensure correct runtime behavior. [[1]](diffhunk://#diff-f0bff30e893c06fbb3d8d69fd03e697b2b0c9a3e8d044accee7338c6ce202300R412-R427) [[2]](diffhunk://#diff-f0bff30e893c06fbb3d8d69fd03e697b2b0c9a3e8d044accee7338c6ce202300R445-R449) [[3]](diffhunk://#diff-f0bff30e893c06fbb3d8d69fd03e697b2b0c9a3e8d044accee7338c6ce202300R555-R579)

**Controller and Documentation Updates:**

* Modified the node health controller to use NodePool-specific repair toleration durations when available, falling back to CloudProvider defaults as needed.
* Added a new design document (`nodepool-repair-configuration.md`) detailing the motivation, API changes, usage examples, and design decisions for the NodePool repair configuration feature. Also updated the main node repair design doc to reference this new capability. [[1]](diffhunk://#diff-4a76e58e2289ae67e3c6ffb449739c423d8d075e2b81f6274a80c410d050cea5R1-R220) [[2]](diffhunk://#diff-997c287391c6ff3ac98a6fada129945da32e2c6de6d9c225352315b207527aeaR129)

engedaam commented on designs/node-repair.md:
```diff
@@ -18,99 +18,145 @@ The alpha implementation will not consider these features:
 The team does not have enough data to determine the right level of configuration that users will utilize. The opinionated mechanism would be responsible for defining unhealthy notes. The advantage of creating the mechanism would be to reduce the configuration burden for customers. **The feature will be gated under an alpha NodeRepair=true feature flag. This will  allow for additional feedback fro ...
 
-## Recommendation: CloudProvider-Defined RepairPolicies
+## Recommendation: CloudProvider-Defined RepairPolicies with NodePool-Configurable TolerationDuration
```
Can you write a separate section in this doc or move the changes you want to make to a separate design? It will make it hard to track in the future all the design choice and consideration we have made

engedaam commented on designs/node-repair.md:
```diff
@@ -18,99 +18,145 @@ The alpha implementation will not consider these features:
+    TolerationDuration *metav1.Duration
+}
+
+type NodePoolSpec struct {
```
Can you describe your use case as to why we need to expose all of these details to users? Also, are you expecting cloudprovider to give default repair policies?

engedaam replied:
Also should this be as part of the disruption stanza?

matanbaruch replied:
Done

engedaam commented on designs/node-repair.md:
```diff
@@ -18,99 +18,145 @@ The alpha implementation will not consider these features:
+    // DefaultTolerationDuration is the default duration to wait before repairing
+    // nodes for any condition not explicitly configured in Policies.
+    // If not specified, uses the CloudProvider's default durations.
+    DefaultTolerationDuration *metav1.Duration
```
Why define this if we give all the toleration duration for every policy? How will the interaction between customer defined polices?

matanbaruch replied:
Fixed.

engedaam commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,29 @@
+{
```
What's the purpose of the devcontonatiners? Why do we need this to merged into the project?

matanbaruch replied:
I had difficulties setting up the environment for testing (golangci-lint version, go versioned installed)

Dev containers is the most efficient way to setup an environment regardless the developer environment.
Also support all of the IDE's.

https://code.visualstudio.com/docs/devcontainers/containers

jonathan-innis commented:
/assign @engedaam

engedaam commented on designs/node-repair.md:
```diff
@@ -22,97 +22,176 @@ The alpha implementation will not consider these features:
 
```
@matanbaruch instead of changing the node repair doc, can you separate out your use cases and proposal into their own section. It might be better to write this as it's own RFC, since this a pretty major change to the NodePool API. Main reason here is that I would good for us to keep the historical context for the choices that were made with the feature.

matanbaruch replied:
You are right. Fixed

matanbaruch commented:
Resolve https://github.com/aws/karpenter-provider-aws/issues/7491

engedaam commented on pkg/cloudprovider/types.go:
```diff
@@ -49,6 +49,14 @@ var (
+}
+
+// RepairPolicy is deprecated, use RepairStatement instead
 type RepairPolicy struct {
```
Why can't drop this struct?

matanbaruch replied:
Fixed. The `RepairStatement` struct has been dropped entirely. Only `RepairPolicy` (the upstream struct with `ConditionType`, `ConditionStatus`, and `TolerationDuration`) remains. The CloudProvider interface is unchanged from upstream — `RepairPolicies() []RepairPolicy`.

engedaam reviewed (commented):
I reviewed the design document, and did an initial pass on the code. Before doing a more through review on the code change, let's close on some of the design questions?

alikhil commented:
Hey @matanbaruch, this is a really great proposal. I'd love to see this feature merged 🚀

Is there anything I can help with to get it over the finish line? Happy to jump in!

matanbaruch commented on designs/nodepool-repair-configuration.md:
```diff
@@ -0,0 +1,198 @@
+- Availability zone resiliency considerations
+- Customer-defined node conditions for repair (beyond CloudProvider-defined conditions)
+
+## Alternative Designs Considered
```
Fixed. The section has been restructured. The current design is now the main proposal in the doc. What was previously mislabeled has been moved to "Alternative Designs Considered" at the bottom, which now correctly documents two rejected alternatives:

1. **Placing Repair Under the Disruption Stanza** — rejected because repair is forceful and doesn't respect disruption budgets/TGP/do-not-disrupt
2. **Allowing User-Defined Conditions** — rejected because CloudProvider should be authoritative for which conditions indicate infrastructure failures

matanbaruch commented on pkg/cloudprovider/types.go:
```diff
@@ -49,6 +49,14 @@ var (
 
 type DriftReason string
 
+type RepairStatement struct {
```
Fixed. `RepairStatement` has been removed. Only `RepairPolicy` exists now — the same struct from upstream with `ConditionType`, `ConditionStatus`, and `TolerationDuration`. No changes to `pkg/cloudprovider/types.go` from upstream.

matanbaruch commented on designs/nodepool-repair-configuration.md:
```diff
@@ -0,0 +1,198 @@
+
+While cloud providers can define default repair policies, different workloads have different sensitivities to node issues:
+
+- **Critical, stateful applications** may require longer toleration durations to allow for manual intervention or graceful failover
```
Great question. The Motivation section now covers this — for stateful applications like databases and distributed storage, automated repair triggered too quickly can disrupt ongoing data replication or cause split-brain scenarios. Operators typically need time for manual verification (e.g., checking replication lag, ensuring quorum) before allowing a node to be replaced. The longer toleration duration gives them that window. This was motivated by real-world experience running stateful workloads on Kubernetes.

matanbaruch commented on .devcontainer/post-create.sh:
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
```
Fixed. Devcontainer files have been removed from this PR.

matanbaruch commented on designs/node-repair.md:
```diff
@@ -125,5 +39,5 @@ There are additional features we will consider including after the initial itera
-* Additional consideration for the availability zone resiliency 
-
+* Additional consideration for the availability zone resiliency
+* Customer-defined node conditions for repair (beyond CloudProvider-defined conditions) 
```
Fixed. The original node-repair.md is no longer modified (historical context preserved). The new RFC is in a separate document.

matanbaruch commented:
@engedaam Ready for a re-review

erezzarum commented:
Does this also support the reason behind the condition (event)?

matanbaruch committed:
feat: Introduce NodePool repair policies with toleration durations

Adds a `repair` stanza to NodePoolSpec so users can override the
CloudProvider's default node-repair TolerationDuration per NodePool
for specific node conditions.

Toleration duration resolves in this order:
1. NodePool policy matching the condition type and status
2. CloudProvider TolerationDuration (unchanged fallback)

Every override is explicit, so a CloudProvider adding a new repair
condition later cannot inherit an unrelated duration.

The CloudProvider interface is unchanged. Includes the RFC in
designs/nodepool-repair-configuration.md.

matanbaruch commented:
Rebased onto latest main, conflicts are resolved and the CLA check is green again.

What changed since the last review:

1. Dropped `defaultTolerationDuration`. Your point in https://github.com/kubernetes-sigs/karpenter/pull/2310#discussion_r2195664282 was right, one duration applied to every policy is the wrong shape. Overrides are explicit per condition now, so a CloudProvider adding a new repair condition later keeps its own duration instead of inheriting an unrelated one.
2. Resolution is down to two steps: a NodePool policy matching `conditionType` and `conditionStatus`, otherwise the CloudProvider's `TolerationDuration`.
3. The CloudProvider interface is still untouched.
4. Squashed to a single commit.

@erezzarum not today. Matching is on `conditionType` and `conditionStatus` only. The condition `Reason` is not part of it, so you can't target a specific reason or the event behind it. If you have a concrete case for that let's do it as a follow up, I'd rather not widen this API further here.

@alikhil thanks for offering. Nothing to split up right now, it's only waiting on review.

@engedaam ready for another look.

matanbaruch committed:
Merge branch 'main' into main
